### PR TITLE
REFACTOR: Metadata object using class

### PR DESCRIPTION
**Pull request by alexandre-merle**
Using a class to compute regular object
metadata

### CODE REVIEW

File: lib/services.js [hunk 1]
```diff
@@ -252,8 +225,11 @@ export default {
                     log.warn('error parsing acl', { error: err });
                     return cb(err);
                 }
-                omVal.acl = parsedACL;
-                metadata.putObjectMD(bucketName, objectKey, omVal, log, err => {
+                md.setAcl(parsedACL);
+                log.trace('object metadata', {
+                    omVal: md.getData(),
```
Comment (electrachong) [inline lib/services.js]: I have heard it's better not to store large objects in the logs to minimize performance and storage impact, not sure if object MD qualifies but it's something to consider

File: lib/services.js [hunk 2]
```diff
@@ -264,7 +240,10 @@ export default {
                 return undefined;
             });
         } else {
-            metadata.putObjectMD(bucketName, objectKey, omVal, log, err => {
+            log.trace('object metadata', {
+                omVal: md.getData(),
```
Comment (electrachong) [inline lib/services.js]: same note here

File: lib/metadata/models/object/object.js [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+        Object.keys(data).forEach(key => { this.data[key] = data[key]; });
+        return this;
+    }
+
+    /**
+     * Returns metadata object
+     *
+     * @return {object} metadata object
+     */
+    getData() {
```
Comment (electrachong) [inline lib/metadata/models/object/object.js]: minor: since we have a public method for accessing the property, you could name the 'data' property '_data' to follow convention, similar to what we do in BucketInfo class

File: lib/services.js (hunk 1 above)
Comment (alexandre-merle) [inline lib/services.js]: you're right :) i have replace what exists before, but we can consider to remove this :) Here its a trace log, so it will be log only if there is an error, no opinion on this one :)

File: lib/metadata/models/object/object.js (hunk 3 above)
Comment (alexandre-merle) [inline lib/metadata/models/object/object.js]: Yes, its better :)

**Comment by alexandre-merle** [comment]
@electrachong yes, here there is only the put of the metadata for regular object, then i have to do the other metdata mpuObject, mpuPart, and then the get will comes later :)

**Comment by alexandre-merle** [comment]
@electrachong updated, unit tests added, we can think in the futur to add more specific check in the setter of the metadata :)

File: lib/services.js (hunk 1 above)
Comment (ThibaultRiviere) [inline lib/services.js]: I think @electrachong is right, even if it's only trace, the logger does some operations like object copy right ?

File: tests/unit/metadata/models/object.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import ObjectMD from '../../../../lib/metadata/models/object/object';
+
+describe('Metadata regular object', () => {
+    let md = null;
+
+    beforeEach(() => {
+        md = new ObjectMD();
+    });
+
+    [
```
Comment (electrachong) [inline tests/unit/metadata/models/object.js]: Maybe add an explanatory comment:
`// In order: data property, value to set/get, default value`

File: tests/unit/metadata/models/object.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
+import assert from 'assert';
+import ObjectMD from '../../../../lib/metadata/models/object/object';
+
+describe('Metadata regular object', () => {
```
Comment (electrachong) [inline tests/unit/metadata/models/object.js]: `describe('ObjectMD class setters/getters', () => {`

File: tests/unit/metadata/models/object.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            FULL_CONTROL: ['id'],
+            WRITE_ACP: ['id'],
+            READ: ['id'],
+            READ_ACP: ['id'],
+        }],
+        ['Key', null, ''],
+        ['Key', 'key'],
+        ['Location', null, []],
+        ['Location', ['location1']],
+    ].forEach(test => {
```
Comment (electrachong) [inline tests/unit/metadata/models/object.js]: even better:
create some variable names and the test will be a lot easier to read:
```
const property = test[0];
const testValue = test[1];
const defaultValue =  test[2]
```

do you agree? :smile:

File: lib/services.js (hunk 1 above)
Comment (alexandre-merle) [inline lib/services.js]: im okay to remove it, but its not my goal, since it can be useful for s3 debugging @LaurenSpiegel ?

File: tests/unit/metadata/models/object.js (hunk 4 above)
Comment (alexandre-merle) [inline tests/unit/metadata/models/object.js]: i can do that ^^

File: tests/unit/metadata/models/object.js (hunk 6 above)
Comment (alexandre-merle) [inline tests/unit/metadata/models/object.js]: take more space on the 80 chars :p joking, i can do that as well ^^

**Comment by alexandre-merle** [comment]
@electrachong updated :)

File: lib/metadata/models/object/object.js [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+        return this._data.location;
+    }
+
+    /**
+     * Set custom headers
+     *
+     * @param {object} data - Headers
+     * @return {ObjectMD} itself
+     */
+    setHeaders(data) {
```
Comment (LaurenSpiegel) [inline lib/metadata/models/object/object.js]: we should do some sanity check to make sure these are all setting x-amz-meta headers.  I do not want to unintentionally overwrite another property like the location.

also, let's call this setUserMetadata

File: lib/services.js (hunk 1 above)
Comment (LaurenSpiegel) [inline lib/services.js]: Let's just keep the log for the acl's.  Everything else should be pretty straightforward.

File: tests/unit/metadata/models/object.js (hunk 6 above)
Comment (electrachong) [inline tests/unit/metadata/models/object.js]: okay, thank you :)

Comment (electrachong) [inline tests/unit/metadata/models/object.js]: but why is it you only test default value for many of them? Do you not want to test all the setter methods, or did I misunderstand?

**Comment by LaurenSpiegel** [comment]
@alexandre-merle, please update so can be merged

File: lib/metadata/models/object/object.js [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+     *
+     * @param {object} acl - Access control list
+     * @param {string} acl.Canned -
+     * @param {string[]} acl.FULL_CONTROL -
+     * @param {string[]} acl.WRITE_ACP -
+     * @param {string[]} acl.READ -
+     * @param {string[]} acl.READ_ACP -
+     * @return {ObjectMD} itself
+     */
+    setAcl(acl) {
```
Comment (rahulreddy) [inline lib/metadata/models/object/object.js]: This needs a class to manage the ACL properties as well. Not in this PR, but definitely needs one.

File: lib/metadata/models/object/object.js [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+    }
+
+    /**
+     * Set location
+     *
+     * @param {string[]} location - location
+     * @return {ObjectMD} itself
+     */
+    setLocation(location) {
+        this._data.location = location;
```
Comment (rahulreddy) [inline lib/metadata/models/object/object.js]: Should clone object to handle array references here.

**Comment by alexandre-merle** [comment]
Updated :)

File: lib/metadata/models/object/object.js [hunk 10]
```diff
@@ -0,0 +1,465 @@
+
+// Version 2 changes the format of the location property
+const modelVersion = 2;
+
+/**
+ * Class to manage metadata object for regular s3 objects
```
Comment (LaurenSpiegel) [inline lib/metadata/models/object/object.js]: please update the comment so it is clear you are distinguishing between the metadata for an mpu part

File: lib/metadata/models/object/object.js [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+     */
+    setLocation(location) {
+        const newLocation = location ? [] : undefined;
+        if (location) {
+            const len = location.length;
+            for (let i = 0; i < len; ++i) {
+                newLocation.push(location[i]);
+            }
+        }
+        this._data.location = location;
```
Comment (LaurenSpiegel) [inline lib/metadata/models/object/object.js]: I think you meant newLocation here.  In any case, each location is an object (other than the legacy version where each location is just a string) so you are just sending references to the the objects into a new array.  I would leave this as it was without the newLocation and just use this._data.location = location

File: lib/metadata/models/object/object.js (hunk 7 above)
Comment (LaurenSpiegel) [inline lib/metadata/models/object/object.js]: not done?

**Comment by LaurenSpiegel** [comment]
@alexandre-merle, it looks like you missed some earlier comments here?

File: lib/metadata/models/object/object.js (hunk 11 above)
Comment (rahulreddy) [inline lib/metadata/models/object/object.js]: I guess this is a problem for some other time.

**Comment by alexandre-merle** [comment]
@LaurenSpiegel indeed yes, i just missed some comments x)

**Comment by alexandre-merle** [comment]
updated :)

File: lib/metadata/models/object/object.js (hunk 7 above)
Comment (LaurenSpiegel) [inline lib/metadata/models/object/object.js]: how is this different from setUserMetadata?  what is the difference?

**Comment by bennettbuchanan** [comment]
Update: Taking over this branch so we can use this way of storing objectMD for bucket replication. 😃

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by bennettbuchanan** [comment]
@ironman-machine try

File: lib/metadata/models/object/ObjectMD.js [hunk 12]
```diff
@@ -0,0 +1,613 @@
...
+
+    /**
+     * Set custom meta headers
+     *
+     * @param {object} data - Meta headers
+     * @return {ObjectMD} itself
+     */
+    setUserMetadata(data) {
+        Object.keys(data).forEach(key => {
+            if (key.startsWith('x-amz-meta')) {
```
Comment (jonathan-gramain) [inline lib/metadata/models/object/ObjectMD.js]: minor: shouldn't the prefix checked against `x-amz-meta-` instead? (with trailing `-`) not sure it really matters but it looks more correct - could suddenly matter e.g. if amazon decided to use some `x-amz-metacool-header` to annoy devs :rofl:

File: lib/metadata/models/object/ObjectMD.js [hunk 13]
```diff
@@ -0,0 +1,613 @@
...
+    }
+
+    /**
+     * Set custom headers
+     *
+     * @param {object} data - Headers
+     * @return {ObjectMD} itself
+     */
+    setHeaders(data) {
+        Object.keys(data).forEach(key => {
```
Comment (jonathan-gramain) [inline lib/metadata/models/object/ObjectMD.js]: May be done as `Object.assign(this._data, data)`?

File: lib/metadata/models/object/ObjectMD.js [hunk 14]
```diff
@@ -0,0 +1,613 @@
...
+     * @return {object} tags object
+     */
+    getTags() {
+        return this._data.tags;
+    }
+
+    /**
+     * Set custom meta headers
+     *
+     * @param {object} data - Meta headers
```
Comment (jonathan-gramain) [inline lib/metadata/models/object/ObjectMD.js]: I suggest renaming `data` to `metaHeaders`

File: lib/metadata/models/object/ObjectMD.js [hunk 15]
```diff
@@ -0,0 +1,613 @@
...
+        return this;
+    }
+
+    /**
+     * Set custom headers
+     *
+     * @param {object} data - Headers
+     * @return {ObjectMD} itself
+     */
+    setHeaders(data) {
```
Comment (jonathan-gramain) [inline lib/metadata/models/object/ObjectMD.js]: I suggest renaming `data` to `headers` or `customHeaders`

File: lib/services.js [hunk 16]
```diff
@@ -95,54 +96,36 @@ const services = {
...
-        // Version 2 changes the format of the location property
-        omVal['md-model-version'] = 2;
+        const md = new ObjectMD();
         // If an IAM user uploads a resource,
         // the owner should be the parent account.
         // http://docs.aws.amazon.com/AmazonS3/
         // latest/dev/access-control-overview.html
-        omVal['owner-display-name'] =
-            authInfo.getAccountDisplayName();
+        md.setOwnerDisplayName(authInfo.getAccountDisplayName());
```
Comment (jonathan-gramain) [inline lib/services.js]: Why not chaining this call too

File: lib/services.js [hunk 17]
```diff
@@ -95,54 +96,36 @@ const services = {
...
-
-        omVal.acl = {
-            Canned: 'private',
-            FULL_CONTROL: [],
-            WRITE_ACP: [],
-            READ: [],
-            READ_ACP: [],
-        };
-
+        if (headers && headers['x-amz-website-redirect-location']) {
```
Comment (jonathan-gramain) [inline lib/services.js]: Saving the website redirect location seems to have been introduced in this commit, maybe deserve a special mention or separate commit (if not separate issue)?

File: lib/metadata/models/object/ObjectMD.js [hunk 18]
```diff
@@ -0,0 +1,613 @@
...
+        });
+        return this;
+    }
+
+    /**
+     * Returns metadata object
+     *
+     * @return {object} metadata object
+     */
+    getData() {
```
Comment (jonathan-gramain) [inline lib/metadata/models/object/ObjectMD.js]: I find the name `getData` confusing, it can be taken for the actual object data. Could be name `getValue` instead for example, or `asJSObject` to point the fact that we're returning the native JS object.

File: tests/unit/metadata/models/object.js [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+        const defaultValue = test[2];
+        const testName = testValue === null ? 'get default' : 'get/set';
+        it(`${testName}: ${property}`, () => {
+            if (testValue !== null) {
+                md[`set${property}`](testValue);
+            }
+            const value = md[`get${property}`]();
+            if ((testValue !== null && typeof testValue === 'object') ||
+                typeof defaultValue === 'object') {
+                assert.deepStrictEqual(value, testValue || defaultValue);
```
Comment (jonathan-gramain) [inline tests/unit/metadata/models/object.js]: Shouldn't you branch the test here depending on if `testValue !== null`? Just to handle the potential case where `testValue === 0`.

File: lib/services.js (hunk 16 above)
Comment (bennettbuchanan) [inline lib/services.js]: I guess this ordering is just left over. I've added it to the chain. 😄

File: lib/metadata/models/object/ObjectMD.js (hunk 13 above)
Comment (bennettbuchanan) [inline lib/metadata/models/object/ObjectMD.js]: That's true, as there are no conditional checks here.

File: lib/metadata/models/object/ObjectMD.js (hunk 15 above)
Comment (bennettbuchanan) [inline lib/metadata/models/object/ObjectMD.js]: Updated.

File: lib/metadata/models/object/ObjectMD.js (hunk 12 above)
Comment (bennettbuchanan) [inline lib/metadata/models/object/ObjectMD.js]: Good point. Updated.

File: lib/metadata/models/object/ObjectMD.js (hunk 14 above)
Comment (bennettbuchanan) [inline lib/metadata/models/object/ObjectMD.js]: Updated.

File: lib/metadata/models/object/ObjectMD.js [hunk 20]
```diff
@@ -0,0 +1,613 @@
...
+     */
+    setVersionId(versionId) {
+        this._data.versionId = versionId;
+        return this;
+    }
+
+    /**
+     * Get metadata versionId value
+     *
+     * @return {boolean} The object versionId
```
Comment (bennettbuchanan) [inline lib/metadata/models/object/ObjectMD.js]: Good catch. Thanks!

File: tests/unit/metadata/models/object.js (hunk 19 above)
Comment (bennettbuchanan) [inline tests/unit/metadata/models/object.js]: Good catch. Updated.

File: lib/services.js (hunk 17 above)
Comment (bennettbuchanan) [inline lib/services.js]: Looks like this objectMD value was originally handled by this loop: https://github.com/scality/S3/pull/582/files#diff-304c00e406caafeefbc9c0a7d27086d8L188. I've added the check to instead include the conditional logic in the setUserMetadata method.

File: lib/metadata/models/object/ObjectMD.js (hunk 18 above)
Comment (bennettbuchanan) [inline lib/metadata/models/object/ObjectMD.js]: Updated.

File: lib/metadata/models/object/ObjectMD.js [hunk 21]
```diff
@@ -0,0 +1,613 @@
...
+    }
+
+    /**
+     * Set location
+     *
+     * @param {string[]} location - location
+     * @return {ObjectMD} itself
+     */
+    setLocation(location) {
+        this._data.location = JSON.parse(JSON.stringify(location));
```
Comment (bennettbuchanan) [inline lib/metadata/models/object/ObjectMD.js]: This was to clone the object. I guess we should be cloning any object properties though.

Comment (jonathan-gramain) [inline lib/metadata/models/object/ObjectMD.js]: An `Object.assign()` instead would not work? (I assume it's because we need a deep copy here)

**Comment by bennettbuchanan** [comment]
@ironman-machine try

File: lib/services.js (hunk 17 above)
Comment (bennettbuchanan) [inline lib/services.js]: After conversation, keeping this as is.

File: lib/metadata/models/object/object.js (hunk 9 above)
Comment (bennettbuchanan) [inline lib/metadata/models/object/object.js]: Keeping this as is due to performance considerations, and since locations should not be modified outside of this class.

File: lib/metadata/models/object/ObjectMD.js [hunk 22]
```diff
@@ -595,7 +595,7 @@ module.exports = class ObjectMD {
      * @param {object} headers - Headers
      * @return {ObjectMD} itself
      */
-    setHeaders(headers) {
+    overrideMetadataValues(headers) {
```
Comment (LaurenSpiegel) [inline lib/metadata/models/object/ObjectMD.js]: please update description of function above to: overrideMetadataValues (used for complete MPU and object copy)

Comment (bennettbuchanan) [inline lib/metadata/models/object/ObjectMD.js]: Updated.

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by bennettbuchanan** [comment]
E2E passed on retry: http://ci.ironmann.io/gh/scality/Integration/13243.
